### PR TITLE
feat: initiate eth withdrawal

### PR DESCRIPTION
**Pull request by GopherJ**
## Security Checklist

- [ ] 1. Re-Entrancy
- [ ] 2. Arithmetic Over/Under Flows
- [ ] 3. Unexpected Ether
- [ ] 4. Delegatecall
- [ ] 5. Default Visibilities
- [ ] 6. Entropy Illusion
- [ ] 7. External Contract Referencing
- [ ] 8. Short Address/Parameter Attack (off chain)
- [ ] 9. Unchecked CALL Return Values
- [ ] 10. Race Conditions / Front Running
- [ ] 11. Denial Of Service (DOS)
- [ ] 12. Block Timestamp Manipulation
- [ ] 13. Constructors with Care
- [ ] 14. Uninitialized Storage Pointers
- [ ] 15. Floating Points and Precision
- [ ] 16. Tx.Origin Authentication
- [ ] 17. Address.isContract Re-Entrancy via Constructor

### ⚠️ NOTES ⚠️

Make sure to think about each of these exploits in this PR.

### CODE REVIEW

File: contracts/misc/ETHWithdrawal.sol [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+        address recipient,
+        uint256 amount
+    ) external nonReentrant {
+        TokenInfo memory tokenInfo = tokenInfos[tokenId];
+        if (tokenInfo.provider == IETHWithdrawal.StakingProvider.Validator) {
+            if (block.timestamp < tokenInfo.withdrawableTime) {
+                revert NotMature();
+            }
+
+            Helpers.safeTransferETH(recipient, tokenInfo.balance);
```
Comment (jfzhou5) [inline contracts/misc/ETHWithdrawal.sol:119]: Helpers.safeTransferETH(recipient, amount) ?

File: contracts/misc/ETHWithdrawal.sol [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+            }
+
+            tokenInfos[tokenId] = IETHWithdrawal.TokenInfo(
+                provider,
+                exitEpoch,
+                withdrawableEpoch,
+                balance,
+                withdrawableTime
+            );
+            _mint(recipient, tokenId, balance, bytes(""));
```
Comment (jfzhou5) [inline contracts/misc/ETHWithdrawal.sol:100]: Is balance used as the amount? Just like the user has 32e18 NFTs with this id?

Comment (GopherJ) [reply contracts/misc/ETHWithdrawal.sol:100]: yes

File: contracts/misc/ETHWithdrawal.sol (hunk 1 above)
Comment (GopherJ) [reply contracts/misc/ETHWithdrawal.sol:119]: emmm you are right

**Pull request merged**